src/vs/workbench/api/node/extHostTypes.ts

```diff
@@ -1626,8 +1626,34 @@ export enum TaskScope {
 	Workspace = 2
 }
 
-export class Task implements vscode.Task {
+export class ExtensionCallbackExecution implements vscode.ExtensionCallbackExecution {
+	private _callback: (args: vscode.TerminalRenderer, cancellationToken: vscode.CancellationToken) => Thenable<void>;
 
+	constructor(callback: (args: vscode.TerminalRenderer, cancellationToken: vscode.CancellationToken) => Thenable<void>) {
+		this._callback = callback;
+	}
+
+	public computeId(): string {
+		const hash = crypto.createHash('md5');
+		hash.update('extensionCallback');
+		// TODO: How is this ID used. Not sure how to create a valid ID from a function.
```

Comment:
@dbaeumer, how is this ID used? I'm also not seeing it. Based on what I can see, using `generateUuid()` should be fine. Since it would be possible to have multiple tasks with the same callback code we can't just use `this._callback.toString()`.

Reply:
TL\DR - The ID is simply used by the extension host and the main thread host to determine if a task has been modified, and hence is considered a different task.

This Todo can actually be removed. The ID really uniquely identifies tasks in the extension host process for any task that was created by an extension. The ID is reset whenever an extension modifies any task properties. I.e. once something is modified, it becomes a different task to both the extension host and the main thread. When the main thread decides to execute a task, it results in $onDidStartTask being called which in turn ends up creating a mapping between the main thread's concept of an ID and the extension hosts concept of an ID by calling getTaskExecution. (Yeah super complicated right? :)) As long as the provided tasks have the same properties, and the same definition, then the IDs will be consistent.

Now, an overall question is that if an extension modifies its properties of a task (lets say with the same names) between calls to provideTasks and there is a task with the same name "running", I don't see any way that the extension host or the main thread host will know that the task returned from a provider is actually conceptually the same task as before. (Have not verified this). It seems that a provider should be able to give a task a unique ID to say "Yep, this really is the same task I provided before regardless of task definition, callback, etc."